Unit tests Improvement

takato1314 opened the pull request:
Add unit tests to the tutor project from 60 tests to 121 tests with a total code coverage of 63.57%. The detailed changes are described in the following pull requests in my committing repo:
- [#1](https://github.com/takato1314/tutor/pull/1)
- [#2](https://github.com/takato1314/tutor/pull/2)
- [#3](https://github.com/takato1314/tutor/pull/3)
- [#4](https://github.com/takato1314/tutor/pull/4)
- [#5](https://github.com/takato1314/tutor/pull/5)

![image](https://user-images.githubusercontent.com/10910968/143991270-3ebb9dce-8ec4-44d1-bac2-3a9767bdd9ac.png)

takato1314 commented:
Btw, I do notice that the one of the latest changes in the current master branch has failed tests on it and I have fixed it on my branch via [#5](https://github.com/takato1314/tutor/pull/5). Was curious if the test pipeline is not working at the moment?

regisb commented on bin/main.py:
```diff
@@ -19,6 +19,5 @@
     except ImportError:
         pass
 
-from tutor.commands.cli import main
```
There is a reason that this import is placed after plugins are loaded. I guess I should have written that down at the time because I can't remember that reason at all... 

Ah I get it! At the time the plugin commands were being added to the cli object during module import: https://github.com/overhangio/tutor/blob/159e24a95d52826f3072261d3b96d3f4247e931a/tutor/commands/cli.py#L64
The plugin commands are now added as part of the `tutor.commands.cli.main()` function call. So your change should be fine and is indeed a good idea :+1:

regisb commented on docs/conf.py:
```diff
@@ -28,7 +28,7 @@
 # -- Sphinx-Click configuration
 # https://sphinx-click.readthedocs.io/
-extensions.append('sphinx_click')
+extensions.append("sphinx_click")
```
We should add the `./docs` to the `SRC_DIRS` in `Makefile`. That way, this file will automatically be linted and formatted in the future. Actually, when I see your changes to `setup.py` below, I'm thinking that we should add this file as well.

takato1314 replied:
There are some mypy errors which I do not know how to resolve when I'm trying to add `./docs` to the `SRC_DIRS` . Is it possible if we add it on a different PR?

takato1314 replied:
I have done that but there will be a few mypy errors which I would need your feedback on how to resolve it afterwards when I checked in the codes. The pipeline will surely report that failed test.

regisb commented on setup.py:
```diff
@@ -21,15 +24,22 @@ def load_about():
+
 def is_requirement(line):
-    return not (line.strip() == "" or line.startswith("#"))
+    return not (line.strip() == "" or line.startswith("#") or line.startswith("-"))
```
I don't understand: which case is this check supposed to cover?

takato1314 replied:
Hi @regisb , when running the unit tests, the setup will load the requirements for unit tests (i.e. `load_dev_requirements()`) via the [`tests_require`](https://github.com/overhangio/tutor/blob/3e61feced4b6b126bddd631e29ba5782fe4a6489/setup.py#L81). This will then cause error when trying to load the requirements because the `dev.in` file contains this [specific line `-r base.txt`](https://github.com/overhangio/tutor/blob/3e61feced4b6b126bddd631e29ba5782fe4a6489/requirements/dev.in#L1), which needs to be ignored by the `tests_require` like the other comments.

takato1314 replied:
I have reverted the changes here back to the original usage without the new load_dev_requirements() function via my repo PR  [#8](https://github.com/takato1314/tutor/pull/8).

regisb commented on tests/commands/test_cli.py:
```diff
@@ -0,0 +1,25 @@
+import unittest
+
+from click.testing import CliRunner
+from tutor.commands.cli import *
```
Please avoid start imports. Instead, import the `cli` module and refer to all objects as attributes of this module. Same comment for all test files below.

takato1314 replied:
@regisb, if I understand you correctly you meant the star/asterisk `*` import on line 4? I already resolved that for now if that's what you meant.

regisb commented on tests/test.py:
```diff
@@ -0,0 +1,33 @@
+from tutor.commands.context import BaseJobContext
+
+
+CONTEXT = Context(os.path.join(tempfile.gettempdir(), "tutor"))
```
This is causing a single context to be used to all unit tests. On my machine this is `/tmp/tutor`. Instead, a different tutor root should be created for all tests, with a randomly generated name so as not to overwrite existing tutor roots. So this global variable should be converted to a `def make_context()` function. Also, this test context should really be temporary and be deleted after each test run.

takato1314 replied:
The main reason I made single static context is to prevent the work needed to scaffold each unique context root and cleanup (i.e. remove) it after the single test is done to keep the test performance time to as minimum as possible. But if that is not a concern for you then I'm find with your idea to make unique context for each test.

takato1314 replied:
The changes has been made into [`helpers.py`](https://github.com/overhangio/tutor/blob/f4f5d6001167341913688f43215b89ac013f84e6/tests/helpers.py) to ensure all TestContext are created with new unique root directory for each test case.

Usage is as follows:
``` python
    def test_load_yml_config(self) -> None:
        with TestContext() as context:
            config = context.load_config()
            current = tutor_config.load(context.root)
            self.assertEqual(config, current)
```

Due to the scaffolding and cleanup for the TestContext() for each test case, the overall unit test time has increased until around 35seconds:
![image](https://user-images.githubusercontent.com/10910968/144938241-c2d07705-3427-4c23-a436-5e9457ed50f4.png)

regisb commented on tests/test.py:
```diff
@@ -0,0 +1,33 @@
+import os
```
This module does not contain any actual unit test, so I suggest it is named `tests/helpers.py`. It should be imported in all test modules as `from .helpers import ...`.

regisb commented on tests/commands/test_cli.py:
```diff
@@ -0,0 +1,25 @@
+        runner = CliRunner()
+        result = runner.invoke(print_help)
+        self.assertEqual(0, result.exit_code)
+        self.assertEqual(None, result.exception)
```
This call and all similar lines should be converted to `self.assertIsNone(result.exception)`.

regisb commented on tests/commands/test_config.py:
```diff
@@ -0,0 +1,92 @@
+        runner = CliRunner()
+        result = runner.invoke(config_command, ["--help"])
+        self.assertEqual(0, result.exit_code)
+        self.assertFalse(result.exception)
```
Is this False or None?

takato1314 replied:
Didn't notice there was this function self.assertIsNone(result.exception). Change to use that instead for every related unit tests scenario.

regisb commented on tests/commands/test_config.py:
```diff
@@ -0,0 +1,92 @@
+        self.assertEqual(0, result.exit_code)
+        self.assertFalse(result.exception)
+        result = runner.invoke(config_command, ["printvalue", "key"], obj=CONTEXT)
+        self.assertNotEqual(-1, result.output.find("value"))
```
We can convert this to `self.assertIn(value, result.output)`.

regisb commented on tests/commands/test_images.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_images_pull_plugin(self) -> None:
+        runner = CliRunner()
+        result = runner.invoke(images_command, ["pull", "plugin"], obj=CONTEXT)
```
`tutor pull plugin` is not an actual command. In this context, `plugin` will be interpreted as an image name, which will not exist. Right now this is not triggering an error, but it will once this other PR is merged: https://github.com/overhangio/tutor/pull/535
So I suggest that we either remove this test or mock the `images.pull` function.

takato1314 replied:
I end up mocking the `images.pull` function.

regisb commented on tests/test_jobs.py:
```diff
@@ -0,0 +1,84 @@
+        service = re.search(r"Service: (\w*)", output)
+        commands = re.search(r"(-----)([\S\s]+)(-----)", output)
+        assert service is not None
+        assert commands is not None
```
We should convert this to `assertIsNone` statements (likewise for other `assert ...` statements below).

takato1314 replied:
@regisb, I tried to use assertIsNone but this will still trigger the error from mypy as shown below.
![image](https://user-images.githubusercontent.com/10910968/144870790-40c7a7db-ae3a-4142-b6d8-3357081d9464.png)

The only way I know at the moment is to use `assert *** is not None` so that [mypy would be able to ascertain that the object is indeed not null.](https://stackoverflow.com/a/68471435/3104587)

regisb commented on tests/commands/test_plugins.py:
```diff
@@ -0,0 +1,45 @@
+        result = runner.invoke(plugins_command, ["list"], obj=CONTEXT)
+        self.assertEqual(0, result.exit_code)
+        self.assertEqual(None, result.exception)
+        self.assertFalse(result.output)
```
This statement is failing on my machine because I have many plugins:

```
======================================================================
FAIL: test_plugins_list (commands.test_plugins.PluginsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/regis/projets/overhang/repos/overhang/tutor/tests/commands/test_plugins.py", line 27, in test_plugins_list
    self.assertFalse(result.output)
AssertionError: 'cairn==12.0.11 (disabled)\ndiscovery==12.0.4 (disabled)\necommerce==12.0.2 (disabled)\nforum==12.0.0 (disabled)\nlicense==12.0.1 (disabled)\nmachine==12.0.1 (disabled)\nmfe==12.1.0 (disabled)\nminio==12.0.1 (disabled)\nmonitor==12.0.2 (disabled)\nnotes==12.0.0 (disabled)\nrichie==12.0.0 (disabled)\nstacktracer==0.1.0 (disabled)\nwebui==12.0.1 (disabled)\nxqueue==12.1.0 (disabled)\n' is not false

----------------------------------------------------------------------
```

So in order for this test to pass we should mock the `plugins.iter_installed` function such that it returns an empty list.

regisb commented on tests/test_utils.py:
```diff
@@ -45,3 +67,94 @@ def test_rsa_key(self) -> None:
+        except KeyboardInterrupt:
+            process.communicate.assert_called_once()
+            process.wait.assert_called_once()
+            process.kill.assert_called_once()
```
We must be entering this exception catch block, right? Then, the test should fail if we don't. We should write something similar to:

        with self.assertRaises(KeyboardInterrupt)
            utils.execute("echo", "")
        process.communicate.assert_called_once()
        process.wait.assert_called_once()
        process.kill.assert_called_once()

regisb commented on tutor/config.py:
```diff
@@ -185,15 +187,13 @@ def convert_json2yml(root: str) -> None:
-            "Both config.json and config.yml exist in {}: only one of these files must exist to continue".format(
-                root
-            )
+            f"Both config.json and {_CONFIG_FILE} exist in {root}: only one of these files must exist to continue"
```
f-string formatting is not supported in Python 3.5, thus we should avoid it. Same for the change below and in tests/test.py.

regisb replied:
OK I just learned understand that Python 3.5 had reached end of life a year ago and that Tutor was not compatible anymore because we depend on astroid 2.8.3 :sweat_smile: Nonetheless, let's not add f-strings for now, and we'll add them after we explicitely drop Python 3.5 compatibility from setup.py.

regisb replied:
Here is the PR that drops Python 3.5 compatibility (in the nightly branch): https://github.com/overhangio/tutor/pull/542

takato1314 replied:
@regisb since you already [bump up the minimum python requirement to 3.6](https://github.com/overhangio/tutor/pull/542/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R55) now, do you still want me to separate all f-string formatting as a separate PR in the future?

regisb replied:
Well, yes because we only dropped the python 3.5 compatibility in the nightly branch, while your PR targets the master branch.

takato1314 replied:
All f-strings are replaced with format strings for this PR.

regisb commented on tutor/config.py:
```diff
@@ -4,6 +4,8 @@
 from . import env, exceptions, fmt, plugins, serialize, utils
 from .types import Config, cast_config
 
+_CONFIG_FILE = "config.yml"
```
If this variable is meant to be imported elsewhere, let's not prefix it with an underscore. Also, it's a filename, so let's name it `CONFIG_FILENAME`.

regisb commented on tutor/utils.py:
```diff
@@ -187,14 +188,15 @@ def is_a_tty() -> bool:
         try:
-            result = p.wait(timeout=None)
+            p.communicate()
+            result = p.returncode
```
This change is making me nervous. I know that I fiddled quite a bit to get it just right. Are you positive that it's not breaking anything? Is there a particular reason you made this change?

takato1314 replied:
Hi @regisb , I particularly changed this because I need to mock one of the methods in the `subprocess.Popen` module (i.e `wait()`) and then inject mock functions to [simulate Exception or KeyboardInterrupt in test_utils.py](https://github.com/takato1314/tutor/blob/3e61feced4b6b126bddd631e29ba5782fe4a6489/tests/test_utils.py#L130). 

However, this is not possible with `wait()` because you are killing the sub-process and gracefully waiting it to shutdown [using the `wait()`](https://github.com/takato1314/tutor/blob/master/tutor/utils.py#L202) as well. So, it's not possible for me to mock the `wait()` method and hence I used the `communicate()` instead and mock that. Afaik, the `communicate()` should also able to run the sub-process and exists gracefully with the tuple returned.
![image](https://user-images.githubusercontent.com/10910968/144875303-c4cafb16-775b-4775-bff3-7cf6465f559b.png)

So far I have tested to run the tutor installation using `tutor local quickstart`, `tutor dev quickstart` and `tutor k8s quickstart` and installed tutor successfully. 

There's already the check at line 210 to check for the fail result and throw error. If you wanna to be extra safe, maybe we could also check the tuple results from the `p.communicate()`?

```
output, error = p.communicate()

## Check error if exist?
if result > 0 or error:
            raise exceptions.TutorError(
                "Command failed with status {}: {}".format(result, " ".join(command))
            )
```

takato1314 replied:
Hi @regisb , fyi idm to revert this part of change as well if you are still worry about it and maybe we can address it in different PR.

takato1314 replied:
Hi @regisb , i reverted thiis part to use the original functions and add in unit tests based on those functions via my repo PR  [#8](https://github.com/takato1314/tutor/pull/8).

regisb reviewed (changes requested):
Hey, thanks a lot for this PR. I understand that it's a lot of work to get started on writing a comprehensive unit test suite, and I appreciate the effort. I added many comments, and I really hope that we can resolve them all because I would very much like to land this PR.

takato1314 commented:
Hi @regisb , thanks for the feedback, I will look at the mentioned comments and adapt them by this weekend.

takato1314 commented on .gitignore:
```diff
@@ -1,12 +1,142 @@
-!.gitignore
-TODO
-__pycache__
+# Copied from https://github.com/github/gitignore/blob/master/Python.gitignore
```
Reverted the entire .gitignore file.

takato1314 commented on .gitignore:
```diff
@@ -1,12 +1,142 @@
+#  Usually these files are written by a python script from a template
+#  before PyInstaller builds the exe, so as to inject date/other infos into it.
+*.manifest
+*.spec
```
Reverted the entire .gitignore file

takato1314 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,45 @@
+{
```
Done excluded from the global gitignore on local dev machine and reverted the entire .gitignore file

takato1314 commented on Makefile:
```diff
@@ -49,6 +49,11 @@ test-pythonpackage: build-pythonpackage ## Test that package can be uploaded to
 	twine check dist/tutor-$(shell make version).tar.gz
 	twine check dist/tutor-openedx-$(shell make version).tar.gz
 
+test-coverage:
```
Hope this is ok for you :)

![image](https://user-images.githubusercontent.com/10910968/144864781-2d52477a-22c5-4d01-975e-a1138d8c9cef.png)

takato1314 commented:
Hi @regisb , please find all of the changes mentioned amended as [in this PR on my own fork](https://github.com/takato1314/tutor/pull/6).

takato1314 commented:
Hi and happy new year @regisb , just updated the PR with the latest master and I also reverted those remaining unresolved comments. 
![image](https://user-images.githubusercontent.com/10910968/147858666-35c0ea43-0e65-4fbb-aca3-5f0edcfc8c15.png)

regisb commented:
I moved my comments to a separate PR: https://github.com/takato1314/tutor/pull/9

takato1314 commented:
Hi @regisb. I've done approving and merging [PR: takato1314#9](https://github.com/takato1314/tutor/pull/9)

regisb commented:
Awesome! Can you please rebase your changes on top of master and squash all your commits? (`git rebase -i master && git push --force`) Then we should be ready to merge.

takato1314 commented:
Hi @regisb , I think I've done it correctly to rebase all changes from this repository. Do let me know if there's any issue.

takato1314 commented:
Please dont approve first. There are some mistakes while rebasing the commits

regisb commented:
Ok let me know if you are facing issues with the rebase. If necessary I can rebase myself.

takato1314 commented:
Hi @regisb. Sry for some delay. Finally figured out the rebase and able to run the tests successfully on my local machine.

takato1314 commented:
Just a quick question, I saw this dependency error while install on windows with the command `pip install -e .`, is this something need to be taken care of? Error doesn't show on my WSL Ubuntu.

takato1314 commented:
Ok I missed the failing error while running `make test` earlier. Will look at it.

regisb commented:
@takato1314 I think that you have multiple errors in your rebase. I wil rebase your changes locally and will push the commit myself if that's ok with you.

takato1314 commented:
Yes please do so I shall close this PR then if that's ok with you?
